# List and Dictionary Comprehensions

Python Comprehension is a different way to construct lists and dictionaries. 

Up to now, every time that we have built a list or dictionary, we did the following:

1. Initializing the list or dictionary and declaring it as a variable.
2. Built them one element/key-value-pair at a time, taking advantage of their inherent mutability.

This is fine, but there is a more succinct way to accomplish the vast majority of your list and dictionary construction tasks.

## Objectives

At the end of this notebook you should be able to:

- Use list comprehension

- Use dictionary comprehension

- Use tuple comprehension

## List Comprehensions

Before we dive into the specifics about how this new tool (list comprehensions) works, let's look at an example question where we build a list.  We can then show how to perform the same task with our new tool and learn how it works.

### Question:

Let's imagine that we have the list `[1, 5, 9, 33]` stored as the variable `my_list`. 

Let's assume that we want to make a new list of the **squares** of all the values in `my_list` and call it `my_squares`. 

With the tools covered so far, you might write:

In [10]:
my_list = [1, 5, 9, 33] # Current list
my_squares = [] # Empty list to store squares

for num in my_list:
    my_squares.append(num ** 2)
    
print(my_list)
print(my_squares)

[1, 5, 9, 33]
[1, 25, 81, 1089]


Now, the `my_squares` list will hold `[1, 25, 81, 1089]`. To get this, we simply specified a bunch of stuff that we wanted to add on to the end of the `my_squares` list, with a starting point at `my_list`. 

So, from a high level, we can code the framework for creating a list as:

```python
list_we_are_building = []
for thing in iterable:
    list_we_are_building.append(transform(thing))
```
With this structure in mind, we can use the following syntax to perform the same task in a single line! Check it out...

```python
list_we_are_building_better = [transform(thing) for thing in iterable]
```

This single line of code does the exact same thing as the three lines above. Lets break this down:

The thing we would pass to the `append()` method, `transform(thing)`, comes at the beginning of the statement inside `[]`.  

**These `[]` allow for the final product to be defined as a list.**

Then, the `for` loop statement is at the end. This is the basic idea behind the [List Comprehension](https://en.wikipedia.org/wiki/List_comprehension).

So, we can now build the `my_squares` list using a list comprehension:

In [ ]:
my_squares_but_better = [num ** 2 for num in my_list]

print(my_squares_but_better)
print(set(my_squares_but_better) == set(my_squares))

[1, 25, 81, 1089]
True


**Remember**: In all the examples where you were getting evens, we had a condition to decide when to append a value to a list? We can also use conditionals to determine what "transformed things" get added in a list comprehension! 

Let's look at the evens list builder.

```python
# Old way of constructing list of evens
evens = []
for num in range(10):
    if num % 2 == 0:
        evens.append(num)

# Old way at high level
list_were_building = []
for thing in iterable:
    if condition:
        list_were_building.append(transform(thing))

# List comprehension way of constructing list of evens
evens = [num for num in range(10) if num % 2 == 0]

# List comprehension way at high level
list_we_are_building = [transform(thing) for thing in iterable if condition]
```

The way `transform()` was called in the above examples can also be considered as if it were a function.

This is an option when writing list comprehensions. For example, a `square` function can accomplish the same result:

```python
# A function that returns the squares of a number
def square(num):
    return num ** 2

# List conprehension that uses the function
my_squares = [square(num) for num in my_list]
```

This might seem silly, since we could just write `num ** 2` directly in the list comp as we did above. However, the calling of a function in the list comp becomes a powerful idea when you want to transform the values being iterated over in a complex way.

For example:

In [22]:
# Function to square numbers but only if they are not divisible by 3.

def square_not_3(num):
    if num % 3 == 0:
        pass
    else:
        return num ** 2
    
my_squares_not_3 = [square_not_3(num) for num in my_list]
print(my_squares_not_3)

[1, 25, None, None]


## Dictionary Comprehensions

Just as list comprehensions are a more succinct way of constructing a list, we have the same ability for dictionaries. 

Dictionary comprehensions operate in the same way as their list counterparts, except for one fundamental difference. 

Recall that dictionaries have no `append()` method, and that a new key-value pair is added to the dictionary with the syntax: 

```python
my_dict[new_key] = new_value
```

In this way, it makes sense that we need syntax to pass both the key and value to the dictionary comprehension.

Luckily, Python gives a simple way to pass a key and value pair, and it is already very familiar to you! You just separate the key and value that you want to enter into the dictionary with a colon `:`, like we did when we were hardcoding the contents in the `{}` dictionary constructor, i.e. 

```python
my_dict = {1: 1, 2: 4}
```

Let's look at an example where we make a dictionary with the keys as the numbers 1 - 5, and the values as the squares of the keys. We'll do this with both the old way of constructing a dictionary and then with a dictionary comprehension so that we can see the similarities.

In [23]:
# The standard way you have seen before
squares_dict = {} # Empty dictionary

for num in range(1, 6): 
    squares_dict[num] = num ** 2

print(squares_dict)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}


In [24]:
# The Dictionary Comprehension method
#  
better_squares_dict = {num: num ** 2 for num in range(1, 6)}
print(better_squares_dict)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25}


We can see that in both cases, we're going through the numbers 1 - 5 with `range(1, 6)` and those `num`s are being assigned as keys. 

The values assigned to those keys are the squares of the keys, assigned with `squares_dict[num] = num ** 2` and `num: num ** 2`, respectively. Just as with list comprehensions, dictionary comprehensions read as the first thing being the `key: value` pair being added to the dictionary. Then, left to right (top down in the old way), we have what the loop definition would look like. 

Just as with list comprehensions, we can add conditionals to filter what gets put into the dictionary.

Say that we want a dictionary with a random integer between 1 and 10, associated with each of the values in the list of words: `['cow', 'chicken', 'horse', 'moose']`. Let's look at how we'd do that with a dictionary comprehension. 

**Note**: We're importing from the Python library `random` to get our random integers. We'll talk more about importing later.

In [51]:
from random import randint
animals_list = ['Cow', 'Chicken', 'Horse', 'Moose']

In [56]:
animals_dict = {animal: randint(1, 10) for animal in animals_list}
print(animals_dict)

{'Cow': 6, 'Chicken': 7, 'Horse': 9, 'Moose': 1}


And we can add functions in a same way as list comprehensions, bulding on complexity.

In [57]:
new_animals_dict = {animal: square_not_3(randint(1, 10)) for animal in animals_list}
print(new_animals_dict)

{'Cow': 4, 'Chicken': None, 'Horse': None, 'Moose': 25}


## Tuple Comprehensions

You can actually use the syntax from the list comprehensions to construct a tuple in a dynamic way. 

Take the example.


In [58]:
my_tuple = tuple(num for num in range(10) if num % 2 == 0)
print(my_tuple)

(0, 2, 4, 6, 8)


All we are doing here is passing `num for num in range(10) if num % 2 == 0` to the tuple constructor, `()`. Since the tuple constructor takes any iterable, which that statement produces, it makes a tuple out of the contents. 

**Note** that it would be impossible to make a tuple with statements like this the "old way", since tuples don't support appending or mutation of any kind.

For this reason, in addition to their readability, comprehensions of all types are considered the most Pythonic way of constructing new data structures.

## Check your understanding

1. Take the following for loop, and translate it into a list. comprehensions: 

```python
odds = []
for num in range(10): 
    if num % 2 != 0: 
        odds.append(num)
```

In [ ]:
# Answer Here

       
2. Take the following for loop, and translate it into a dictionary comprehensions: 

```python
cubes = {}
for num in range(1, 6): 
    cubes[num] = num **  3
```

In [61]:
# Answer Here